tick/eval/eval_binary_node.go

```diff
@@ -0,0 +1,334 @@
+	if err != nil {
+		return nil, fmt.Errorf("Failed to handle left node: %v", err)
+	}
+
+	rightSideEvaluator, err := createNodeEvaluator(node.Right)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to handle right node: %v", err)
+	}
+
+	expression.leftSideEvaluator = leftSideEvaluator
+	expression.rightSideEvaluator = rightSideEvaluator
+
+	if isDynamicNode(node.Left) || isDynamicNode(node.Right) {
+		expression.evaluationFn = expression.evaluateDynamicNode
+	} else {
+		expression.leftSideType = getConstantNodeType(node.Left)
+		expression.rightSideType = getConstantNodeType(node.Right)
+
+		expression.evaluationFn = expression.findEvaluationFn(expression.leftSideType, expression.rightSideType)
+	}
+
+	return expression, nil
+}
+
+func (n *EvalBinaryNode) GetType(scope ReadOnlyScope, executionState ExecutionState) (ValueType, error) {
+	return findNodeTypes(n.Type, []NodeEvaluator{n.leftSideEvaluator, n.rightSideEvaluator}, scope, executionState)
+}
+
+func (n *EvalBinaryNode) EvalRegex(scope *tick.Scope, executionState ExecutionState) (*regexp.Regexp, error) {
+	return nil, CreateTypeGuardFailedError(TRegex, n.Type)
+}
+
+func (n *EvalBinaryNode) EvalString(scope *tick.Scope, executionState ExecutionState) (string, error) {
+	return "", CreateTypeGuardFailedError(TString, n.Type)
+}
+
+// EvalBool executes the expression based on eval bool
+func (e *EvalBinaryNode) EvalBool(scope *tick.Scope, executionState ExecutionState) (bool, error) {
+	result, err := e.evalWithNodes(scope, executionState, e.leftSideEvaluator, e.rightSideEvaluator)
+	if err != nil {
+		return false, err.error
+	}
+
+	if result.IsBoolValue {
+		return result.BoolValue, nil
+	}
+
+	return false, fmt.Errorf("expression returned unexpected type %T", result.value())
+}
+
+// EvalNum executes the expression based on eval numeric
+func (e *EvalBinaryNode) EvalFloat64(scope *tick.Scope, executionState ExecutionState) (float64, error) {
+	result, err := e.evalWithNodes(scope, executionState, e.leftSideEvaluator, e.rightSideEvaluator)
+	if err != nil {
+		return float64(0), err.error
+	}
+
+	if result.IsFloat64Value {
+		return result.Float64Value, nil
+	}
+
+	if result.IsInt64Value {
+		return float64(0), CreateTypeGuardFailedError(TFloat64, TInt64)
+	}
+
+	return float64(0), CreateTypeGuardFailedError(TFloat64, e.Type)
+}
+
+func (e *EvalBinaryNode) EvalInt64(scope *tick.Scope, executionState ExecutionState) (int64, error) {
+	result, err := e.evalWithNodes(scope, executionState, e.leftSideEvaluator, e.rightSideEvaluator)
+	if err != nil {
+		return int64(0), err.error
+	}
+
+	if result.IsInt64Value {
+		return result.Int64Value, nil
+	}
+
+	if result.IsFloat64Value {
+		return int64(0), CreateTypeGuardFailedError(TInt64, TFloat64)
+	}
+
+	return int64(0), CreateTypeGuardFailedError(TInt64, e.Type)
+}
+
+func (e *EvalBinaryNode) evalWithNodes(scope *tick.Scope, executionState ExecutionState, left, right NodeEvaluator) (resultContainer, *ErrSide) {
```

Comment:
As far as I can see `evalWithNodes` is always called with the same nodes i.e. e.leftSideEvaluator and e.rightSideEvaluator. Should we drop the  extra arguments and rename it to `evaluate`? 

That was the binary node has two `eval` methods:

* evaluate -- expects to be able to evaluate by directly calling the evaualtionFn
* evaluateDynamic -- first figures out what the evaluationFn should be then calls evaluate.

Does that work?